### PR TITLE
[2405 Rebase & FF] Add additional features to PCD enumeration and only have the PCI Bus master enabled when the device is connected.

### PR DESCRIPTION
**Pull request by kenlautner**
## Description
Commits from the 202311 branch:
09382fe
6b4c06c
ff6324a

The commits do the following:

---

Only turn on the PCI Bus Master Enable when the device is connected, and disable PCI Bus Master Enable at Exit Boot Services.  These features are controlled by Dynamic PCD's.

---

Add support for initializing Max Payload Size during PCIe enumeration.

---

Change assert to not fire when intentionally disabling PCI enumeration

---

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Passes CI

## Integration Instructions

N/A

### CODE REVIEW

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c [hunk 1]
```diff
@@ -15,6 +15,83 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 //
 LIST_ENTRY  mPciDevicePool;
 
+// MU_CHANGE Begin - Disable BME at Exit Boot Services
+
+/**
+ * DisableBmeOnTree
+ *
+ * @param Bridge
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: This parameter really should be updated to match the actual parameter name with a description.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c [hunk 2]
```diff
@@ -15,6 +15,83 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+{
+  LIST_ENTRY     *CurrentLink;
+  PCI_IO_DEVICE  *PciIoDevice;
+  UINT16         Command;
+
+  // DisableBME on ExitBootServices should be synchonized with VTd disable - with
+  // DisableBME running before VTd disable.  If the VTd disable code runs at TPL_CALLBACK,
+  // then DisableBME will run before VTd disable.
+  if (!PcdGetBool (PcdDisableBMEonEBS)) {
+    DEBUG ((EFI_D_WARN, "%a PCI Bus Master enabled due to PCD setting\n", __FUNCTION__));
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Recommendation.

```suggestion
DEBUG ((DEBUG_WARN, "%a PCI Bus Master enabled due to PCD setting\n", __FUNCTION__));
```

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c [hunk 3]
```diff
@@ -24,7 +101,25 @@ InitializePciDevicePool (
...
+  Status = gBS->CreateEventEx (
+                  EVT_NOTIFY_SIGNAL,
+                  TPL_NOTIFY,
+                  OnExitBootServices,
+                  NULL,
+                  &gEfiEventExitBootServicesGuid,
+                  &ExitBootServicesEvent
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Unable to create ExitBootServices event. COde=%r\n", Status));
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]:
```suggestion
DEBUG ((DEBUG_ERROR, "Unable to create ExitBootServices event. Code=%r\n", Status));
```

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c (hunk 2 above)
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Recommendation (feel free to ignore) is to update to `__func__` while adding changes.

```suggestion
DEBUG ((EFI_D_WARN, "%a PCI Bus Master enabled due to PCD setting\n", __func__));
```

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c [hunk 4]
```diff
@@ -684,7 +779,17 @@ StartPciDevicesOnBridge (
...
-        Supports &= (UINT64)EFI_PCI_DEVICE_ENABLE;
+
+        // MU_CHANGE [Begin] - Defer BME enablement until PCI SetAttributes
+        if (FeaturePcdGet (PcdDeferBME)) {
+          Supports &= (UINT64)(EFI_PCI_IO_ATTRIBUTE_IO | EFI_PCI_IO_ATTRIBUTE_MEMORY);
+        } else {
+          Supports &= (UINT64)EFI_PCI_DEVICE_ENABLE;
+        }
+
+        // Supports &= (UINT64)EFI_PCI_DEVICE_ENABLE;
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Drop the commented line of code.

```suggestion

```

File: MdeModulePkg/MdeModulePkg.dec [hunk 5]
```diff
@@ -954,12 +954,23 @@
...
+  #   FALSE - Initialize P2P bridges with BME enabled.
+  # @Prompt Defer BME enablement
+  gEfiMdeModulePkgTokenSpaceGuid.PcdDeferBME|TRUE|BOOLEAN|0x30002050
+
+[PcdsFeatureFlag.IA32, PcdsFeatureFlag.ARM, PcdsFeatureFlag.AARCH64]
   gEfiMdeModulePkgTokenSpaceGuid.PcdPciDegradeResourceForOptionRom|FALSE|BOOLEAN|0x0001003a
 
 [PcdsFeatureFlag.IA32, PcdsFeatureFlag.X64]
   ## Indicates if DxeIpl should switch to long mode to enter DXE phase.
+  #  It is assumed that 64-bit DxeCore is built in firmware if it is true; otherwise 32-bit DxeCore
```
Comment (makubacki) [inline MdeModulePkg/MdeModulePkg.dec]: This seems unrelated to the other changes in the commit.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h [hunk 6]
```diff
@@ -284,6 +284,8 @@ struct _PCI_IO_DEVICE {
   UINT16                                       BridgeIoAlignment;
   UINT32                                       ResizableBarOffset;
   UINT32                                       ResizableBarNumber;
+  BOOLEAN                                      IgnoreROM;      // MS_CHANGE
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h]: Recommend making these `MU_CHANGE` throughout code modifications for consistency.

**Review by makubacki — APPROVED** [review]

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h:1]: The PR title and description needs to be updated to be more accurate to the changes being made. Currently it is "Add support for PCI enumeration" but PCI enumeration is already supported so its unclear what is actually being added.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c (hunk 1 above)
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Updated to include description of the function and parameter.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c (hunk 2 above)
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Fixed.

Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Fixed.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c (hunk 4 above)
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Removed.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h (hunk 6 above)
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h]: Removed.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c (hunk 3 above)
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciDeviceSupport.c]: Fixed

File: MdeModulePkg/MdeModulePkg.dec (hunk 5 above)
Comment (kenlautner) [reply MdeModulePkg/MdeModulePkg.dec]: Removed.

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h
Comment (kenlautner) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h:1]: I updated the PR title to be accurate (it was previously used from one of the cherry-picked commits).   Do you think it's fine now?

Comment (makubacki) [reply MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.h:1]: Yes, thanks

File: MdeModulePkg/MdeModulePkg.dec [hunk 7]
```diff
@@ -1656,7 +1665,6 @@
   ## Indicates the default timeout value for SD/MMC Host Controller operations in microseconds.
   # @Prompt SD/MMC Host Controller Operations Timeout (us).
   gEfiMdeModulePkgTokenSpaceGuid.PcdSdMmcGenericTimeoutValue|1000000|UINT32|0x00000031
-
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec]: We shouldn't remove this

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/MdeModulePkg.dec (hunk 7 above)
Comment (kenlautner) [reply MdeModulePkg/MdeModulePkg.dec]: Fixed.

**Pull request merged**